plat/common/arm/smp.c

```diff
@@ -0,0 +1,428 @@
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <arm/smp.h>
+#include <arm/smp_defs.h>
+#include <arm/arm64/cpu.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+#include <uk/plat/irq.h>
+#include <arm/time.h>
+#include <uk/plat/time.h>
+#include <arm/smccc.h>
+#include <gic/gic.h>
+
+/**
+ * The number of cells for the size field should be 0 in cpu nodes.
+ * The number of cells in the address field is set by default to 2 in cpu
+ * nodes. This is also the maximum number of cells for the address field that
+ * we can currently support.
+ **/
+#define FDT_SIZE_CELLS_DEFAULT 0
+#define FDT_ADDR_CELLS_DEFAULT 2
+#define CPU_ID_MASK 0xff00ffffffUL
+
+struct arm_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static unsigned int smp_numcores;
+static __lcpuid ukplat_lcpu_idx(void);
+static void *dtb;
+
+void lcpu_start(struct arm_cpu *cpu);
+void __noreturn lcpu_entry_default(void);
+extern struct _gic_dev *gic;
+
+
+static int arm64_lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct ukplat_lcpu_func *i;
+	struct uk_list_head *fnlist;
+
+	__lcpuid this_cpu_idx = ukplat_lcpu_idx();
+
+	cpus[this_cpu_idx].state = LCPU_STATE_BUSY;
+	ukarch_spin_unlock(&cpus[this_cpu_idx].lock);
+
+	fnlist = &cpus[this_cpu_idx].fnlist;
+
+	uk_list_for_each_entry(i, fnlist, lentry) {
+		i->fn(regs, i);
+		if (i == uk_list_first_entry(fnlist,
+				struct ukplat_lcpu_func, lentry))
+			break;
+	}
+
+	return 1;
+}
+
+static int arm64_lcpu_ipi_wakeup_handler(void *args __unused,
+			struct __regs *regs)
+{
+	return 1;
+}
+
+int smp_init(void *plat_dtb)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *nsize_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, nsize = 0, cell;
+	__lcpuid cpu_idx;
+	int subnode;
+	char bsp_found = 0;
+	__u64 bsp_cpu_id;
+
+	dtb = plat_dtb;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
```

Comment:
Posted this to another comment, but since it's marked as outdated, I post here too so that it doesn't get lost:

If you're saying that the bootcore can also receive IPIs then you might want to initializes it's `fnlist` here too. Right now if you issue an SGI from another core to core zero, Unikraft will crash.